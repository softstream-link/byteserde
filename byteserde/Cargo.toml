[package]
name = "byteserde"
version = "0.1.1"
edition = "2021"
authors = ["Softstream <softstream.link@gmail.com>"]
license-file = "../LICENSE"
description = "A generic serialization/deserialization framework for mapping byte streams to/from arbitrary struct types with focus on performance."
repository = "https://github.com/softstream-link/byteserde"
readme = "../readme.md"
keywords = ["serde", "bytes", "serialization", "derive"]
categories = ["encoding"]
rust-version = "1.69"


[dev-dependencies]
text-diff = "0.4"
env_logger = "0.10"
log = "0.4"
