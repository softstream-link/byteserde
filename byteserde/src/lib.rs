// public
pub mod des_bytes;
pub mod des_slice;
pub mod error;
pub mod prelude;
pub mod ser_heap;
pub mod ser_stack;
pub mod size;
pub mod utils;

// test only
#[cfg(test)]
pub mod unittest;
