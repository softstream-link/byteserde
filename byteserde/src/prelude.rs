pub use super::des_slice::{ByteDeserializeSlice, ByteDeserializerSlice, from_serializer_stack, from_serializer_heap, from_slice};
pub use super::des_bytes::{ByteDeserializeBytes, ByteDeserializerBytes, from_bytes};
// pub use super::error::Result;
pub use super::error::SerDesError;
pub use super::ser_heap::{to_bytes_heap, to_serializer_heap};
pub use super::ser_heap::{ByteSerializeHeap, ByteSerializerHeap};
pub use super::ser_stack::{to_bytes_stack, to_serializer_stack};
pub use super::ser_stack::{ByteSerializeStack, ByteSerializerStack};
pub use super::size::{ByteSerializedLenOf, ByteSerializedSizeOf};