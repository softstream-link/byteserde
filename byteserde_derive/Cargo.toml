[package]
name = "byteserde_derive"
version = "0.1.0"
edition = "2021"
authors = ["Softstream <softstream.link@gmail.com>"]
license-file = "../LICENSE"
description = "A procedural macro for mapping byte streams to/from arbitrary struct types with focus on performance"
repository = "https://github.com/softstream-link/byteserde"
readme = "../readme.md"
keywords = [ "serde", "bytes", "serialization", "derive" ]
categories = ["enchoding"]
rust-version = "1.69"
[lib]
proc-macro = true

[dependencies]
syn = { version = "2.0", features = ["full"] }
quote = "1.0"

[dev-dependencies]
trybuild = { version = "1.0", features = ["diff"] }
