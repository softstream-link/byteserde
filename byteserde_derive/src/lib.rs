use common::{get_crate_name, get_generics, get_struct_ser_des_tokens};
use proc_macro::TokenStream;
use quote::quote;
use syn::DeriveInput;

mod common;
mod struct_shared;

#[proc_macro_derive(ByteSerializeStack, attributes(byteserde))]
#[allow(non_snake_case)] // keep snake name otherwise it messes up vscode refactoring
pub fn ByteSerializeStack(input: TokenStream) -> TokenStream {
    let ast: DeriveInput = syn::parse(input).unwrap();

    // get struct name
    let struct_name = &ast.ident;
    let (generics_declaration, generics_alias) = get_generics(&ast.generics);
    let (ser_method, _) = get_struct_ser_des_tokens(&ast);
    // grap just stack presets
    let ser_vars = ser_method.iter().map(|f| &f.ser_vars).collect::<Vec<_>>();
    let ser_over = ser_method.iter().map(|f| &f.ser_over).collect::<Vec<_>>();
    let ser_uses_stack = ser_method
        .iter()
        .map(|f| &f.ser_uses_stack)
        .collect::<Vec<_>>();

    let crate_name = get_crate_name();

    // generate stack serializer
    let output = quote! {
        #[automatically_derived]
        impl #generics_declaration #crate_name::ser::ByteSerializeStack for #struct_name #generics_alias{
        // impl byteserde::ser::ByteSerializeStack for #struct_name {
            fn byte_serialize_stack<const CAP: usize>(&self, ser: &mut #crate_name::ser::ByteSerializerStack<CAP>) -> #crate_name::error::Result<()>{
                // numerics
                //      ser.serialize_[be|le|ne](self.field_name)?; -- for regular
                //      ser.serialize_[be|le|ne](self.0         )?; -- for tuple
                // trait ByteSerializeStack
                //      self.field_name.byte_serialize_stack(ser)?;     -- for regular
                //      self.0         .byte_serialize_stack(ser)?;     -- for tuple
                #( #ser_vars )*
                #( #ser_over )*
                #( #ser_uses_stack )*
                Ok(())
            }
        }
    };
    output.into()
}

#[proc_macro_derive(ByteSerializeHeap, attributes(byteserde))]
#[allow(non_snake_case)] // keep snake name otherwise it messes up vscode refactoring
pub fn ByteSerializeHeap(input: TokenStream) -> TokenStream {
    let ast: DeriveInput = syn::parse(input).unwrap();
    // get struct name
    let struct_name = &ast.ident;
    let (generics_declaration, generics_alias) = get_generics(&ast.generics);
    // get ser & des quote presets
    let (ser_method, _) = get_struct_ser_des_tokens(&ast);
    // grap just heap presets
    let ser_vars = ser_method.iter().map(|f| &f.ser_vars).collect::<Vec<_>>();
    let ser_over = ser_method.iter().map(|f| &f.ser_over).collect::<Vec<_>>();
    let ser_uses_heap = ser_method
        .iter()
        .map(|f| &f.ser_uses_heap)
        .collect::<Vec<_>>();

    let crate_name = get_crate_name();

    // generate heap serializer
    let output = quote! {
        #[automatically_derived]
        impl #generics_declaration #crate_name::ser::ByteSerializeHeap for #struct_name #generics_alias{
            fn byte_serialize_heap(&self, ser: &mut #crate_name::ser::ByteSerializerHeap) -> #crate_name::error::Result<()>{
                // numerics
                //      ser.serialize_[be|le|ne](self.field_name)?;         -- for regular
                //      ser.serialize_[be|le|ne](self.0         )?;         -- for tuple
                // trait ByteSerializeStack
                //      self.field_name.byte_serialize_heap(ser)?;          -- for regular
                //      self.0         .byte_serialize_heap(ser)?;          -- for tuple
                #( #ser_vars)*
                #( #ser_over)*
                #( #ser_uses_heap)*
                Ok(())
            }
        }
    };
    output.into()
}

#[proc_macro_derive(ByteDeserialize, attributes(byteserde))]
#[allow(non_snake_case)] // keep snake name otherwise it messes up vscode refactoring
pub fn ByteDeserialize(input: TokenStream) -> TokenStream {
    let ast: DeriveInput = syn::parse(input).unwrap();
    // get struct name
    let struct_name = &ast.ident;
    let (generics_declaration, generics_alias) = get_generics(&ast.generics);
    // get ser & des quote presets
    let (des_method, ty) = get_struct_ser_des_tokens(&ast);
    // grap just heap presets
    let des_vars = des_method.iter().map(|f| &f.des_vars).collect::<Vec<_>>();
    let des_uses = des_method.iter().map(|f| &f.des_uses).collect::<Vec<_>>();
    let impl_body = match ty {
        common::StructType::Regular => quote!(#struct_name {#( #des_uses )*}),
        common::StructType::Tuple => quote!  (#struct_name (#( #des_uses )*)),
    };

    let crate_name = get_crate_name();

    // generate deserializer
    let output = quote! {
        #[automatically_derived]
        impl #generics_declaration #crate_name::des::ByteDeserialize<#struct_name #generics_alias> for #struct_name #generics_alias {
            fn byte_deserialize(des: &mut #crate_name::des::ByteDeserializer) -> #crate_name::error::Result<#struct_name #generics_alias>{
                // let type_u16:    u16 = des.deserialize_[be|le|ne]()?; -- numerics
                // let type_String: String = des.deserialize()?;          -- trait ByteDeserialize
                // StructName { type_u16, type_String }
                //
                // let _0 = des.deserialize_[be|le|ne]()?; -- numerics
                // let _1  = des.deserialize()?;          -- trait ByteDeserialize
                // TupleName ( _0, _1 )
                #( #des_vars )*
                Ok(#impl_body)
            }
        }
    };
    output.into()
}
