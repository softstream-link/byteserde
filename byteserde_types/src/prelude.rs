pub use super::strings::ascii::StringAsciiFixed;
pub use super::strings::ascii::StringAscii;
pub use super::strings::ascii::CharAscii;
pub use super::strings::ascii::ConstCharAscii;
