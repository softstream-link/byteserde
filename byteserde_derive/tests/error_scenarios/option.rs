#[allow(unused_imports)]
use byteserde::prelude::*;

use byteserde_derive::ByteDeserializeSlice;

#[derive(ByteDeserializeSlice)]
struct TupleStruct(u8, u8);

#[derive(ByteDeserializeSlice)]
struct OptionalSectionAllMustBeStruct {
    field: Option<u8>,
    field1: Option<TupleStruct>,
}

#[derive(ByteDeserializeSlice)]
struct OptionalSectionMissingEqAnnotation {
    field1: Option<TupleStruct>,
}

#[derive(ByteDeserializeSlice)]
struct OptionalSectionMissingPeekAnnotation {
    #[byteserde(eq( 1_u16.to_be_bytes() ))]
    field1: Option<TupleStruct>,
}

#[derive(ByteDeserializeSlice)]
#[byteserde(peek(0, 2))]
struct OptionalSectionAllMustBeOption {
    field: TupleStruct,
    #[byteserde(eq( 1_u16.to_be_bytes() ))]
    field1: Option<TupleStruct>,
}

fn main() {}
