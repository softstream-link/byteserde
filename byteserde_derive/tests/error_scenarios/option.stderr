error: proc-macro derive panicked
 --> tests/error_scenarios/option.rs:9:10
  |
9 | #[derive(ByteDeserializeSlice)]
  |          ^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: Option of Byte & Numerics are not supported only of other struct types. Ex: Option<SomeStruct>

error: proc-macro derive panicked
  --> tests/error_scenarios/option.rs:15:10
   |
15 | #[derive(ByteDeserializeSlice)]
   |          ^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: struct `OptionalSectionMissingEqAnnotation` ByteDeserialize error:
           OptionalSectionMissingEqAnnotation.field1 is Option<T> type and hence requires `#[byteserde(eq( ... ))] attribute it that evalutes to a byte slice and complared with &[u8] of `#[byteserde(peek( start, len ))]` expression

error: proc-macro derive panicked
  --> tests/error_scenarios/option.rs:20:10
   |
20 | #[derive(ByteDeserializeSlice)]
   |          ^^^^^^^^^^^^^^^^^^^^
   |
   = help: message: struct `OptionalSectionMissingPeekAnnotation` missing required `#[byteserde(peek( start, len ))]` annotation to be able to identify which optional fields are present in the bytestream
